pkg/globalmanager/controllers/jointinference/jointinferenceservice.go

```diff
@@ -573,5 +626,115 @@ func New(cc *runtime.ControllerContext) (runtime.FeatureControllerI, error) {
 	jc.podStoreSynced = podInformer.Informer().HasSynced
 
+	deploymentInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    jc.addDeployment,
+		UpdateFunc: jc.updateDeployment,
+		DeleteFunc: jc.deleteDeployment,
+	})
+
+	jc.deploymentsLister = deploymentInformer.Lister()
+	jc.deploymentStoreSynced = deploymentInformer.Informer().HasSynced
+
 	return jc, nil
 }
+
+func (c *Controller) addDeployment(obj interface{}) {
+	deployment := obj.(*appsv1.Deployment)
+	c.enqueueByDeployment(deployment)
+}
+
+// deleteDeployment enqueues the JointInferenceService obj When a deleteDeployment is deleted
+func (c *Controller) deleteDeployment(obj interface{}) {
+	deployment, ok := obj.(*appsv1.Deployment)
+
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Warningf("couldn't get object from tombstone %+v", obj)
+			return
+		}
+		deployment, ok = tombstone.Obj.(*appsv1.Deployment)
+		if !ok {
+			klog.Warningf("tombstone contained object that is not a Deployment %+v", obj)
+			return
+		}
+	}
+	c.enqueueByDeployment(deployment)
+}
+
+func (c *Controller) updateDeployment(old, cur interface{}) {
+	oldDeployment := old.(*appsv1.Deployment)
+	curDeployment := cur.(*appsv1.Deployment)
+	// no deployment update, no queue
+	if curDeployment.ResourceVersion == oldDeployment.ResourceVersion {
+		return
+	}
+	c.addDeployment(curDeployment)
+}
+
+func (c *Controller) enqueueByDeployment(deployment *appsv1.Deployment) {
+	controllerRef := metav1.GetControllerOf(deployment)
+
+	klog.Infof("Deployment enqueued %v", deployment.Kind)
+
+	if controllerRef == nil {
+		return
+	}
+
+	if controllerRef.Kind != Kind.Kind {
+		return
+	}
+
+	service, err := c.serviceLister.JointInferenceServices(deployment.Namespace).Get(deployment.Name)
+	if err != nil {
+		return
+	}
+
+	if service.UID != controllerRef.UID {
+		return
+	}
+
+	c.enqueueController(service, true)
+}
+
+func (c *Controller) updateService(old, cur interface{}) {
+	oldService, ok := old.(*sednav1.JointInferenceService)
+	if !ok {
+		return
+	}
+	curService, ok := cur.(*sednav1.JointInferenceService)
+	if !ok {
+		return
+	}
+
+	if oldService == curService {
+		return
+	}
+
+	if reflect.DeepEqual(oldService.Spec, curService.Spec) {
+		return
+	}
+	// if CRD is changed,and the service.Generation is changed, update deployment settings
+	klog.Infof("Service is updated, delete previous deployments")
+	curService.SetGroupVersionKind(Kind)
+	// if the service.Generation is changed, update deployment settings
+	if oldService.Generation != curService.Generation {
+		// delete previous deployments
+		deployments, err := c.deploymentsLister.Deployments(curService.Namespace).List(c.selector)
+		if err != nil {
+			klog.Errorf("Failed to list deployments: %v", err)
+			return
+		}
+		for _, deployment := range deployments {
+			c.kubeClient.AppsV1().Deployments(curService.Namespace).Delete(context.TODO(), deployment.Name, metav1.DeleteOptions{})
```

Comment:
Build deployment first, and then use update function to update, rather than simply delete deployment 
newDeployment := ... 
c.kubeClient.AppsV1().Deployments(newDeployment.Namespace).Update(context.TODO(), newDeployment, metav1.UpdateOptions{})